pypst/utils.py

```diff
@@ -62,3 +101,153 @@ def render_sequence(arg: Iterable[Any]) -> str:
+        '#(bar: 4, qux: none)'
+    """
+
+    def __str__(self) -> str:
+        return self.render()
+
+    def render(self) -> str:
+        """
+        Dataclass rendering to a Typst dictionary.
+
+        Modify `fields_to_render` to control which fields are included.
+        """
+        mapping = render_mapping(
+            {field.name: getattr(self, field.name) for field in self.fields_to_render()}
+        )
+        return f"#{mapping}"
+
+    def fields_to_render(self) -> Iterable[Field]:
+        """
+        These fields should be rendered. Defaults to skipping any fields with a value of `None`.
+        """
+
+        def check(field: Field) -> bool:
+            # When the value is `None`, check whether that should be kept.
+            if getattr(self, field.name) is None:
+                return field.metadata.get("keep_none", False)
+
+            # Keep otherwise.
+            return True
+
+        return filter(check, fields(self))
+
+
+@dataclass
+class Function(Dictionary):
+    """
+    Helper class to render Typst function calls from a Python dataclass.
+
+    Inherit from `Function` to inherit the render method.
+    The function name is derived from the class name
+    and is converted to kebab-case.
+
+    The dataclass' fields are used as the function arguments.
+
+    You can specify a positional argument in Typst by adding
+    `positional=True` on the field's metadata.
+
+    You can specify that `None` values should be rendered instead
+    of being skipped with `keep_none=True` on the field's metadata.
+
+    Example:
+        >>> from dataclasses import dataclass, field
+        >>> @dataclass
+        ... class FooFn(Function):
+        ...    bar: int | None = field(metadata={"positional": True})
+        ...    qux: int | None = field(default=16, metadata={"keep_none": True})
+        >>>
+        >>> FooFn(4).render()
+        '#foo-fn(4, qux: 16)'
+        >>> FooFn(None).render()
+        '#foo-fn(qux: 16)'
+        >>> FooFn(4, qux=None).render()
+        '#foo-fn(4, qux: none)'
+    """
+
+    def render(self) -> str:
+        """
+        Dataclass rendering to a Typst function call.
+
+        Modify `fields_to_render` to control which fields are included.
+        """
+
+        function = self.function()
+
+        options = ", ".join(
+            render_code(getattr(self, field.name))
+            if field.metadata.get("positional", False)
+            else f"{field.name}: {render_code(getattr(self, field.name))}"
+            for field in self.fields_to_render()
+        )
+
+        return f"#{function}({options})"
+
+    @classmethod
+    def function(cls) -> str:
+        """
+        Class method that returns the Typst function name that it represents.
+
+        Defaults to a kebab-case transformation of the ClassName.
+        """
+        return re.sub(
+            r"([a-z0-9])([A-Z])",
+            r"\1-\2",
+            cls.__name__,
+        ).lower()
```

Comment:
Again, nice addition, but I don't think we need the inheritance from `Dictionary`. You only want to use `fields_to_render` which can be its own function.

We can keep this utility class, though, I'd like to propose an alternative. We could introduce a `__is_function__` member for dataclasses that should be rendered as such. It is optional (default `False`) and `boolean`. This would alleviate the need to inherit. Then rendering can be incorporated into the `render` function, too. As I'm on the edge which option would be better, I'd leave the decision up to you.